aries_cloudagent/protocols/did_rotate/v1_0/manager.py

```diff
@@ -0,0 +1,244 @@
+        """
+        async with self.profile.session() as session:
+            record = await RotateRecord.retrieve_by_thread_id(session, ack._thread_id)
+
+        record.state = RotateRecord.STATE_ACK_RECEIVED
+        if not record.new_did:
+            raise ValueError("No new DID stored in record")
+
+        conn.my_did = record.new_did
+        async with self.profile.session() as session:
+            # Don't emit a connection event for this change
+            # Controllers should listen for the rotate event instead
+            await conn.save(session, reason="My DID rotated", event=False)
+            await record.save(session, reason="Received rotate ack")
+
+        # TODO it would be better if the cache key included DIDs so we don't
+        # have to manually clear it. This is a bigger change than a first pass
+        # warrants though.
+        conn_mgr = BaseConnectionManager(self.profile)
+        await conn_mgr.clear_connection_targets_cache(conn.connection_id)
+
+    async def receive_problem_report(
+        self, conn: ConnRecord, problem_report: RotateProblemReport
+    ):
+        """Receive problem report message.
+
+        Args:
+            conn (ConnRecord): The connection to rotate the DID for.
+            problem_report (ProblemReport): The received problem report message.
+        """
+        async with self.profile.session() as session:
+            record = await RotateRecord.retrieve_by_thread_id(
+                session, problem_report._thread_id
+            )
+
+        record.state = RotateRecord.STATE_FAILED
+        # Base ProblemReportSchema requires this value be present
+        assert problem_report.description
+        record.error = problem_report.description["code"]
+        async with self.profile.session() as session:
+            await record.save(session, reason="Received problem report")
+
+    async def receive_hangup(self, conn: ConnRecord, hangup: Hangup):
+        """Receive hangup message.
+
+        Args:
+            conn (ConnRecord): The connection to rotate the DID for.
+            hangup (Hangup): The received hangup message.
+        """
+        # TODO: Should the connection be terminated here?
+
+    async def _ensure_supported_did(self, did: str):
+        """Check if the DID is supported."""
+        resolver = self.profile.inject(DIDResolver)
+        conn_mgr = BaseConnectionManager(self.profile)
+        try:
+            await resolver.resolve(self.profile, did)
+        except DIDMethodNotSupported:
+            raise UnsupportedDIDMethodError(RotateProblemReport.method_unsupported(did))
+        except DIDNotFound:
+            raise UnresolvableDIDError(RotateProblemReport.unresolvable(did))
+
+        try:
+            await conn_mgr.resolve_didcomm_services(did)
+        except BaseConnectionManagerError:
+            # TODO Make this reportable?
+            raise DIDRotateManagerError("Unable to resolve DIDComm services for DID")
```

Comment:
I think we should make this a reportable error

Reply:
I think all three errors get reported back the to the sender at line: https://github.com/petridishdev/aries-cloudagent-python/blob/db05f1e072272f937fd71d886e56179fd136b2fe/aries_cloudagent/protocols/did_rotate/v1_0/manager.py#L123

Correct me if I'm wrong here

Reply:
Actually that one error extends from a non-reported exception. I can update that.

Reply:
The `DIDRotateManagerError` is not a subclass of `ReportableDIDRotateError` so it's not captured in that `try/except` block. It would bubble up to the controller through the Admin API but the sender would not be notified. We would need to add a method like `method_unsupported` on the `RotateProblemReport` message to make this a reportable error.